### ISSUE TITLE
generated defaults lose type information for aliases

### ISSUE THREAD

**Issue by thockin** [kubernetes/gengo#235]
xref #232 

We have many enum-string fields in our APIs, for example:

```
type ServiceInternalTrafficPolicyType string

const (
    ServiceInternalTrafficPolicyCluster ServiceInternalTrafficPolicyType = "Cluster"
    ServiceInternalTrafficPolicyLocal ServiceInternalTrafficPolicyType = "Local"
)
```

I tried to use this as:

```
    // +default=ref(ServiceInternalTrafficPolicyCluster)
    InternalTrafficPolicy *ServiceInternalTrafficPolicyType 
```

The code that got generated looks like:

```
    if in.Spec.InternalTrafficPolicy == nil {
        var ptrVar1 string = ServiceInternalTrafficPolicyCluster
        in.Spec.InternalTrafficPolicy = &ptrVar1                                                                                                     
    }    
```

The problem is that `ServiceInternalTrafficPolicyCluster` is not type `string`, so it fails to compile.

I see in gengo/examples/defaulter-gen/generators/defaulter.go `callTreeForType.build()` it takes types of kind `Alias` and unpacks them, recursively, but that just can't work - we have to carry the alias type name into the generated code.

The fix is not obvious to me and I am out of time to track it for today.

**Comment by nikhita** [comment]
/assign

created #239

**Comment by thockin** [comment]
This one is still open.  The simple fix:

```
diff --git a/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go b/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go
index 48566aee542..ff0b634a0e3 100644
--- a/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go
+++ b/vendor/k8s.io/gengo/examples/defaulter-gen/generators/defaulter.go
@@ -1031,7 +1033,7 @@ func (n *callNode) writeDefaulter(varName string, index string, isVarPointer boo
                // keep using the address-of operator on the primitive value until the types match
                if n.defaultDepth > 0 {
                        sw.Do(fmt.Sprintf("if %s == nil {\n", variablePlaceholder), args)
-                       sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
+                       sw.Do("ptrVar$.varDepth$ := $.defaultValue$\n", args)
                        // We iterate until a depth of 1 instead of 0 because the following line
                        // `if $.varName$ == &ptrVar1` accounts for 1 level already
                        for i := n.defaultDepth; i > 1; i-- {
```

doesn't work because:

```
$ make kube-apiserver
go version go1.20.3 linux/amd64
+++ [0421 13:06:45] Setting GOMAXPROCS: 6
+++ [0421 13:06:45] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kube-apiserver (static)
# k8s.io/kubernetes/pkg/apis/core/v1
pkg/apis/core/v1/zz_generated.defaults.go:234:49: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:574:43: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:707:49: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:1047:52: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:1078:22: cannot use &ptrVar1 (value of type *int) as *int32 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:1135:50: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
pkg/apis/core/v1/zz_generated.defaults.go:1475:58: cannot use &ptrVar1 (value of type *int) as *int64 value in assignment
```

**Comment by apelisse** [comment]
/close

Fixed by #246
